internal/xds/cache/snapshotcache.go

```diff
@@ -0,0 +1,271 @@
+// It does this by also implementing callbacks to make sure that the cache is kept
+// up to date for each new node.
+//
+// Having the cache also implement the callbacks is a little bit hacky, but it makes sure
+// that all the required bookkeeping happens.
+// TODO(youngnick): Talk to the go-control-plane maintainers and see if we can upstream
+// this in a better way.
+type SnapshotCacheWithCallbacks interface {
+	envoy_cache_v3.SnapshotCache
+	envoy_server_v3.Callbacks
+	GenerateNewSnapshot(types.XdsResources) error
+}
+
+type snapshotcache struct {
+	envoy_cache_v3.SnapshotCache
+
+	lastSnapshot *envoy_cache_v3.Snapshot
+
+	log *log.LogrWrapper
+
+	streamIDNodeID map[int64]string
+
+	snapshotVersion int64
+}
+
+// GenerateNewSnapshot takes a table of resources (the output from the IR->xDS
+// translator) and updates the snapshot version.
+func (s *snapshotcache) GenerateNewSnapshot(resources types.XdsResources) error {
+
+	version := s.newSnapshotVersion()
+
+	// Create a snapshot with all xDS resources.
+	snapshot, err := envoy_cache_v3.NewSnapshot(
+		version,
+		resources,
+	)
+	if err != nil {
+		return err
+	}
+
+	s.lastSnapshot = snapshot
+
+	for _, node := range s.getNodeIDs() {
+		fmt.Printf("Generating a snapshot with Node %s\n", node)
+		err := s.SetSnapshot(context.TODO(), node, snapshot)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+// newSnapshotVersion increments the current snapshotVersion
+// and returns as a string.
+func (s *snapshotcache) newSnapshotVersion() string {
+
+	// Reset the snapshotVersion if it ever hits max size.
+	if s.snapshotVersion == math.MaxInt64 {
+		s.snapshotVersion = 0
+	}
+
+	// Increment the snapshot version & return as string.
+	s.snapshotVersion++
+	return strconv.FormatInt(s.snapshotVersion, 10)
+}
+
+// NewSnapshotCache gives you a fresh SnapshotCache.
+// It needs a logger that supports the go-control-plane
+// required interface (Debugf, Infof, Warnf, and Errorf).
+func NewSnapshotCache(ads bool, logger *log.LogrWrapper) SnapshotCacheWithCallbacks {
+	return &snapshotcache{
+		SnapshotCache:  envoy_cache_v3.NewSnapshotCache(ads, &Hash, logger),
+		log:            logger,
+		streamIDNodeID: make(map[int64]string),
+	}
+}
+
+func (s *snapshotcache) getNodeIDs() []string {
+
+	var nodeIDs []string
+
+	for _, node := range s.streamIDNodeID {
+
+		nodeIDs = append(nodeIDs, node)
+
+	}
+
+	return nodeIDs
+
+}
+
+// OnStreamOpen and the other OnStream* functions implement the callbacks for the
+// state-of-the-world stream types.
+func (s *snapshotcache) OnStreamOpen(ctx context.Context, streamID int64, typeURL string) error {
+
+	s.streamIDNodeID[streamID] = ""
+
+	return nil
+}
+
+func (s *snapshotcache) OnStreamClosed(streamID int64) {
+
+	// nodes := s.streamIDNodeID[streamID]
+
+	delete(s.streamIDNodeID, streamID)
+
+}
+
+func (s *snapshotcache) OnStreamRequest(streamID int64, req *envoy_service_discovery_v3.DiscoveryRequest) error {
```

Comment:
curious why SoTW callbacks are implemented, we should just log and return imho

Reply:
Not implementing the full protocol when we can pretty easily makes me uncomfortable - if we do happen to make an error with the Delta callbacks or implementation, the SOTW is a good fallback to have available for operational purposes.

I don't think that not meeting expectations about what variants we support is worth the small saving in development effort.